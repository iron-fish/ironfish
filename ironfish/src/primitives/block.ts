/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { zip } from 'lodash'
import { Assert } from '../assert'
import { BlockHeader, BlockHeaderSerde, SerializedBlockHeader } from './blockheader'
import { MintDescription } from './mintDescription'
import { NoteEncrypted, NoteEncryptedHash } from './noteEncrypted'
import { Nullifier } from './nullifier'
import { SerializedTransaction, Transaction } from './transaction'

/**
 * The hash used in the "previousHash" field on the initial block in the
 * chain. The initial block is intentionally invalid, so we need to special
 * case it.
 */
export const GENESIS_BLOCK_PREVIOUS = Buffer.alloc(32)

export const GENESIS_BLOCK_SEQUENCE = 1

export const GRAFFITI_SIZE = 32

/**
 * Represent a single block in the chain. Essentially just a block header
 * and the list of transactions that were added to the tree between the
 * previous block and the ones committed to in this header.
 */
export class Block {
  header: BlockHeader
  transactions: Transaction[]

  constructor(header: BlockHeader, transactions: Transaction[]) {
    this.header = header
    this.transactions = transactions
  }

  /**
   * Get the number of notes and nullifiers stored on this block.
   */
  counts(): SerializedCounts {
    let notes = 0
    let nullifiers = 0

    for (const transaction of this.transactions) {
      notes += transaction.notes.length
      nullifiers += transaction.spends.length
    }

    return { notes, nullifiers }
  }

  /**
   * Get a list of all spends on transactions in this block.
   *
   * Note: there is no spend on a miner's fee transaction in the header
   */
  *spends(): Generator<{
    nullifier: Nullifier
    commitment: NoteEncryptedHash
    size: number
  }> {
    for (const transaction of this.transactions) {
      for (const spend of transaction.spends) {
        yield spend
      }
    }
  }

  /**
   * Get a list of all notes created in this block including the miner's fee
   * note on the header.
   */
  *notes(): Generator<NoteEncrypted> {
    for (const transaction of this.transactions) {
      for (const note of transaction.notes) {
        yield note
      }
    }
  }

  /**
   * Get a list of all mints on transactions in this block.
   */
  *mints(): Generator<MintDescription> {
    for (const transaction of this.transactions) {
      for (const mint of transaction.mints) {
        yield mint
      }
    }
  }

  equals(block: Block): boolean {
    if (block === this) {
      return true
    }

    if (!this.header.equals(block.header)) {
      return false
    }

    if (this.transactions.length !== block.transactions.length) {
      return false
    }

    for (const [transaction1, transaction2] of zip(this.transactions, block.transactions)) {
      if (!transaction1 || !transaction2 || !transaction1.equals(transaction2)) {
        return false
      }
    }

    return true
  }

  get minersFee(): Transaction {
    const tx = this.transactions[0]
    Assert.isNotUndefined(tx, 'Block has no miners fee')
    return tx
  }

  toCompactBlock(): CompactBlock {
    const header = this.header

    const [minersFee, ...transactions] = this.transactions
    const transactionHashes = transactions.map((t) => t.hash())

    return {
      header,
      transactionHashes,
      transactions: [
        {
          index: 0,
          transaction: minersFee,
        },
      ],
    }
  }
}

export type CompactBlockTransaction = {
  index: number
  transaction: Transaction
}

export type CompactBlock = {
  header: BlockHeader
  transactionHashes: Buffer[]
  transactions: CompactBlockTransaction[]
}

export type SerializedBlock = {
  header: SerializedBlockHeader
  transactions: SerializedTransaction[]
}

export type SerializedCounts = { notes: number; nullifiers: number }

export class BlockSerde {
  static serialize(block: Block): SerializedBlock {
    return {
      header: BlockHeaderSerde.serialize(block.header),
      transactions: block.transactions.map((t) => t.serialize()),
    }
  }

  static deserialize(data: SerializedBlock): Block {
    if (
      typeof data === 'object' &&
      data !== null &&
      'header' in data &&
      'transactions' in data &&
      Array.isArray(data.transactions)
    ) {
      const header = BlockHeaderSerde.deserialize(data.header)
      const transactions = data.transactions.map((t) => new Transaction(t))
      return new Block(header, transactions)
    }
    throw new Error('Unable to deserialize')
  }
}
