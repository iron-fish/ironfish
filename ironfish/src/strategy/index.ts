/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import {
  generateNewPublicAddress,
  WasmNote,
  WasmNoteEncrypted,
  WasmTransactionPosted,
} from 'ironfish-wasm-nodejs'
import { BlockSyncer } from '../blockSyncer'
import {
  Block,
  BlockHash,
  BlockSerde,
  BlockHeader,
  NullifierHasher,
  SerializedBlock,
  Blockchain,
} from '../blockchain'
import { MerkleHasher, Witness } from '../merkletree'
import { Transaction, Spend } from './transaction'
import Serde from '../serde'
import { MiningDirector } from '../mining'
import hashBlockHeader from '../mining/miningAlgorithm'
import { AsyncTransactionWorkerPool } from './asyncTransactionWorkerPool'
import { MemPool } from '../memPool'
import {
  Validity,
  VerificationResult,
  Verifier,
  VerificationResultReason,
} from '../consensus/verifier'

export { Transaction, Spend } from './transaction'
export { default as Strategy } from './strategy'
import Strategy from './strategy'
import { WorkerPool } from '../workerPool'
/**
 * Implementation of the IronFish strategy that calls into sapling via Wasm
 * to encode notes in zero-knowledge proofs.
 */

/**
 * Serialized version of an encrypted note.
 */
export type SerializedWasmNoteEncrypted = Buffer

/**
 * An encrypted note's hash.
 */
export type WasmNoteEncryptedHash = Buffer

/**
 * Serialized version of an encrypted note's hash.
 */
export type SerializedWasmNoteEncryptedHash = Buffer

export const GENESIS_SUPPLY_IN_IRON = 42000000

export const IRON_FISH_YEAR_IN_BLOCKS = 2100000

/**
 * Serde implementation to convert an encrypted note to its serialized form and back.
 */
class WasmNoteEncryptedSerde
  implements Serde<IronfishNoteEncrypted, SerializedWasmNoteEncrypted> {
  equals(note1: IronfishNoteEncrypted, note2: IronfishNoteEncrypted): boolean {
    return note1.serialize().equals(note2.serialize())
  }

  serialize(note: IronfishNoteEncrypted): SerializedWasmNoteEncrypted {
    return note.serialize()
  }

  deserialize(serializedNote: SerializedWasmNoteEncrypted): IronfishNoteEncrypted {
    return new IronfishNoteEncrypted(serializedNote)
  }
}

/**
 * Serde implementation to convert an encrypted note's hash to its serialized form and back.
 */
class WasmNoteEncryptedHashSerde
  implements Serde<WasmNoteEncryptedHash, SerializedWasmNoteEncryptedHash> {
  equals(hash1: WasmNoteEncryptedHash, hash2: WasmNoteEncryptedHash): boolean {
    return hash1.equals(hash2)
  }
  serialize(note: WasmNoteEncryptedHash): SerializedWasmNoteEncryptedHash {
    return note
  }
  deserialize(serializedNote: SerializedWasmNoteEncryptedHash): WasmNoteEncryptedHash {
    return serializedNote
  }
}

/**
 * Hasher implementation for notes to satisfy the MerkleTree requirements.
 */
export class NoteHasher
  implements
    MerkleHasher<
      IronfishNoteEncrypted,
      WasmNoteEncryptedHash,
      SerializedWasmNoteEncrypted,
      SerializedWasmNoteEncryptedHash
    > {
  _merkleNoteSerde: WasmNoteEncryptedSerde
  _merkleNoteHashSerde: WasmNoteEncryptedHashSerde
  constructor() {
    this._merkleNoteSerde = new WasmNoteEncryptedSerde()
    this._merkleNoteHashSerde = new WasmNoteEncryptedHashSerde()
  }

  elementSerde(): Serde<IronfishNoteEncrypted, SerializedWasmNoteEncrypted> {
    return this._merkleNoteSerde
  }

  hashSerde(): Serde<WasmNoteEncryptedHash, SerializedWasmNoteEncryptedHash> {
    return this._merkleNoteHashSerde
  }

  merkleHash(note: IronfishNoteEncrypted): Buffer {
    return note.merkleHash()
  }

  combineHash(
    depth: number,
    left: WasmNoteEncryptedHash,
    right: WasmNoteEncryptedHash,
  ): WasmNoteEncryptedHash {
    return Buffer.from(WasmNoteEncrypted.combineHash(depth, left, right))
  }
}

export type TransactionHash = Buffer

export class IronfishNote {
  private readonly wasmNoteSerialized: Buffer
  private wasmNote: WasmNote | null = null
  private referenceCount = 0

  constructor(wasmNoteSerialized: Buffer) {
    this.wasmNoteSerialized = wasmNoteSerialized
  }

  serialize(): Buffer {
    return this.wasmNoteSerialized
  }

  takeReference(): WasmNote {
    this.referenceCount++
    if (this.wasmNote === null) {
      this.wasmNote = WasmNote.deserialize(this.wasmNoteSerialized)
    }
    return this.wasmNote
  }

  returnReference(): void {
    this.referenceCount--
    if (this.referenceCount <= 0) {
      this.referenceCount = 0
      this.wasmNote?.free()
      this.wasmNote = null
    }
  }

  value(): BigInt {
    const value = this.takeReference().value
    this.returnReference()
    return value
  }

  memo(): string {
    const memo = this.takeReference().memo
    this.returnReference()
    return memo
  }

  nullifier(ownerPrivateKey: string, position: BigInt): Buffer {
    const buf = Buffer.from(this.takeReference().nullifier(ownerPrivateKey, position))
    this.returnReference()
    return buf
  }
}

export class IronfishNoteEncrypted {
  private readonly wasmNoteEncryptedSerialized: Buffer
  private wasmNoteEncrypted: WasmNoteEncrypted | null = null
  private referenceCount = 0

  constructor(wasmNoteEncryptedSerialized: Buffer) {
    this.wasmNoteEncryptedSerialized = wasmNoteEncryptedSerialized
  }

  serialize(): Buffer {
    return this.wasmNoteEncryptedSerialized
  }

  takeReference(): WasmNoteEncrypted {
    this.referenceCount++
    if (this.wasmNoteEncrypted === null) {
      this.wasmNoteEncrypted = WasmNoteEncrypted.deserialize(this.wasmNoteEncryptedSerialized)
    }
    return this.wasmNoteEncrypted
  }

  returnReference(): void {
    this.referenceCount--
    if (this.referenceCount <= 0) {
      this.referenceCount = 0
      this.wasmNoteEncrypted?.free()
      this.wasmNoteEncrypted = null
    }
  }

  decryptNoteForOwner(ownerHexKey: string): IronfishNote | undefined {
    const note = this.takeReference().decryptNoteForOwner(ownerHexKey)
    this.returnReference()
    if (note) {
      const serializedNote = note.serialize()
      note.free()
      return new IronfishNote(Buffer.from(serializedNote))
    }
  }

  decryptNoteForSpender(spenderHexKey: string): IronfishNote | undefined {
    const note = this.takeReference().decryptNoteForSpender(spenderHexKey)
    this.returnReference()
    if (note) {
      const serializedNote = note.serialize()
      note.free()
      return new IronfishNote(Buffer.from(serializedNote))
    }
  }

  merkleHash(): Buffer {
    const note = this.takeReference().merkleHash()
    this.returnReference()
    return Buffer.from(note)
  }
}

/**
 * Wraps a Wasm transaction to provide a convenient interface.
 *
 * Transactions come from a client looking to spend. They are stored on blocks
 * in the transaction list, and one is also used to hold the miner's fee for
 * a given transaction.
 */
export class IronfishTransaction
  implements Transaction<IronfishNoteEncrypted, WasmNoteEncryptedHash> {
  private readonly wasmTransactionPostedSerialized: Buffer
  private readonly workerPool: WorkerPool

  private wasmTransactionPosted: WasmTransactionPosted | null = null
  private referenceCount = 0

  constructor(wasmTransactionPostedSerialized: Buffer, workerPool: WorkerPool) {
    this.wasmTransactionPostedSerialized = wasmTransactionPostedSerialized
    this.workerPool = workerPool
  }

  serialize(): Buffer {
    return this.wasmTransactionPostedSerialized
  }

  takeReference(): WasmTransactionPosted {
    this.referenceCount++
    if (this.wasmTransactionPosted === null) {
      this.wasmTransactionPosted = WasmTransactionPosted.deserialize(
        this.wasmTransactionPostedSerialized,
      )
    }
    return this.wasmTransactionPosted
  }

  returnReference(): void {
    this.referenceCount--
    if (this.referenceCount <= 0) {
      this.referenceCount = 0
      this.wasmTransactionPosted?.free()
      this.wasmTransactionPosted = null
    }
  }

  withReference<R>(callback: (transaction: WasmTransactionPosted) => R): R {
    const transaction = this.takeReference()
    try {
      return callback(transaction)
    } finally {
      this.returnReference()
    }
  }

  async verify(): Promise<VerificationResult> {
    const result = await this.workerPool.verify(this)
    return result === true
      ? { valid: Validity.Yes }
      : { valid: Validity.No, reason: VerificationResultReason.ERROR }
  }

  notesLength(): number {
    return this.withReference((t) => t.notesLength)
  }

  getNote(index: number): IronfishNoteEncrypted {
    return this.withReference((t) => {
      // Get the note
      const serializedNote = Buffer.from(t.getNote(index))

      // Convert it to an IronfishNoteEncrypted
      return new IronfishNoteEncrypted(serializedNote)
    })
  }

  *notes(): Iterable<IronfishNoteEncrypted> {
    const notesLength = this.notesLength()

    for (let i = 0; i < notesLength; i++) {
      yield this.getNote(i)
    }
  }

  spendsLength(): number {
    return this.withReference((t) => t.spendsLength)
  }

  *spends(): Iterable<Spend<WasmNoteEncryptedHash>> {
    const spendsLength = this.spendsLength()
    for (let i = 0; i < spendsLength; i++) {
      yield this.withReference((t) => {
        const wasmSpend = t.getSpend(i)
        const spend: Spend<WasmNoteEncryptedHash> = {
          size: wasmSpend.treeSize,
          nullifier: Buffer.from(wasmSpend.nullifier),
          commitment: Buffer.from(wasmSpend.rootHash),
        }
        wasmSpend.free()
        return spend
      })
    }
  }

  transactionFee(): Promise<bigint> {
    return this.workerPool.transactionFee(this)
  }

  transactionSignature(): Buffer {
    return this.withReference((t) => Buffer.from(t.transactionSignature))
  }

  transactionHash(): TransactionHash {
    return this.withReference((t) => Buffer.from(t.transactionHash))
  }
}

/**
 * Serialized version of the Transaction wrapper.
 */
export type SerializedTransaction = Buffer

export type SerializedIronfishBlock = SerializedBlock<
  SerializedWasmNoteEncryptedHash,
  SerializedTransaction
>

/**
 * Serializer and equality checker for Transaction wrappers.
 */
export class TransactionSerde implements Serde<IronfishTransaction, SerializedTransaction> {
  constructor(private readonly workerPool: WorkerPool) {}

  equals(): boolean {
    throw new Error(`Not implemented`)
  }

  serialize(transaction: IronfishTransaction): SerializedTransaction {
    return transaction.serialize()
  }

  deserialize(data: SerializedTransaction): IronfishTransaction {
    return new IronfishTransaction(data, this.workerPool)
  }
}

/**
 * Implementation of a Blockchain Strategy using the Wasm zero-knowledge proofs.
 */
export class IronfishStrategy
  implements
    Strategy<
      IronfishNoteEncrypted,
      WasmNoteEncryptedHash,
      IronfishTransaction,
      SerializedWasmNoteEncrypted,
      SerializedWasmNoteEncryptedHash,
      SerializedTransaction
    > {
  _noteHasher: NoteHasher

  _nullifierHasher: NullifierHasher

  _blockSerde: BlockSerde<
    IronfishNoteEncrypted,
    WasmNoteEncryptedHash,
    IronfishTransaction,
    SerializedWasmNoteEncrypted,
    SerializedWasmNoteEncryptedHash,
    SerializedTransaction
  >
  private _verifierClass: typeof IronfishVerifier
  private miningRewardCachedByYear: Map<number, number>
  private readonly workerPool: WorkerPool

  constructor(workerPool: WorkerPool, verifierClass: typeof IronfishVerifier | null = null) {
    this._noteHasher = new NoteHasher()
    this._nullifierHasher = new NullifierHasher()
    this._blockSerde = new BlockSerde(this)
    this._verifierClass = verifierClass || Verifier
    this.miningRewardCachedByYear = new Map<number, number>()
    this.workerPool = workerPool
  }

  noteHasher(): NoteHasher {
    return this._noteHasher
  }

  nullifierHasher(): NullifierHasher {
    return this._nullifierHasher
  }

  transactionSerde(): TransactionSerde {
    return new TransactionSerde(this.workerPool)
  }

  get blockSerde(): BlockSerde<
    IronfishNoteEncrypted,
    WasmNoteEncryptedHash,
    IronfishTransaction,
    SerializedWasmNoteEncrypted,
    SerializedWasmNoteEncryptedHash,
    SerializedTransaction
  > {
    return this._blockSerde
  }

  hashBlockHeader(serializedHeader: Buffer): BlockHash {
    return hashBlockHeader(serializedHeader)
  }

  /**
   * See https://ironfish.network/docs/whitepaper/4_mining#include-the-miner-reward-based-on-coin-emission-schedule
   *
   * Annual coin issuance from mining goes down every year. Year is defined here by the
   * number of blocks (IRON_FISH_YEAR_IN_BLOCKS)
   *
   * Given the genesis block supply (GENESIS_SUPPLY_IN_IRON) the formula to calculate
   * reward per block is:
   * (genesisSupply / 4) * e ^(-.05 * yearsAfterLaunch)
   * Where e is the natural number e (Euler's number), and -.05 is a decay function constant
   *
   * @param sequence Block sequence
   * @returns mining reward (in ORE) per block given the block sequence
   */
  miningReward(sequence: bigint): number {
    const yearsAfterLaunch = Math.floor(Number(sequence) / IRON_FISH_YEAR_IN_BLOCKS)
    let reward = this.miningRewardCachedByYear.get(yearsAfterLaunch)
    if (reward) {
      return reward
    }

    const annualReward = (GENESIS_SUPPLY_IN_IRON / 4) * Math.E ** (-0.05 * yearsAfterLaunch)
    reward = this.convertIronToOre(annualReward / IRON_FISH_YEAR_IN_BLOCKS)
    this.miningRewardCachedByYear.set(yearsAfterLaunch, reward)

    return reward
  }

  convertIronToOre(iron: number): number {
    return Math.round(iron * 10 ** 8)
  }

  createVerifier(chain: IronfishBlockchain): IronfishVerifier {
    return new this._verifierClass(chain)
  }

  async createMinersFee(
    totalTransactionFees: bigint,
    blockSequence: bigint,
    minerKey: string,
  ): Promise<IronfishTransaction> {
    const transaction = AsyncTransactionWorkerPool.createTransaction()

    // Generate a public address from the miner's spending key
    const owner = generateNewPublicAddress(minerKey).public_address

    // Create a new note with value equal to the inverse of the sum of the
    // transaction fees and the mining reward
    const amount = totalTransactionFees + BigInt(this.miningReward(blockSequence))
    const minerNote = new WasmNote(owner, amount, '')
    const serializedNote = Buffer.from(minerNote.serialize())
    minerNote.free()

    await transaction.receive(minerKey, new IronfishNote(serializedNote))

    return new IronfishTransaction(
      Buffer.from((await transaction.postMinersFee(this.workerPool)).serialize()),
      this.workerPool,
    )
  }
}

export type IronfishBlock = Block<
  IronfishNoteEncrypted,
  WasmNoteEncryptedHash,
  IronfishTransaction,
  SerializedWasmNoteEncrypted,
  SerializedWasmNoteEncryptedHash,
  SerializedTransaction
>

export type IronfishBlockHeader = BlockHeader<
  IronfishNoteEncrypted,
  WasmNoteEncryptedHash,
  IronfishTransaction,
  SerializedWasmNoteEncrypted,
  SerializedWasmNoteEncryptedHash,
  SerializedTransaction
>

export type IronfishBlockchain = Blockchain<
  IronfishNoteEncrypted,
  WasmNoteEncryptedHash,
  IronfishTransaction,
  SerializedWasmNoteEncrypted,
  SerializedWasmNoteEncryptedHash,
  SerializedTransaction
>

export type IronfishMemPool = MemPool<
  IronfishNoteEncrypted,
  WasmNoteEncryptedHash,
  IronfishTransaction,
  SerializedWasmNoteEncrypted,
  SerializedWasmNoteEncryptedHash,
  SerializedTransaction
>

export class IronfishVerifier extends Verifier<
  IronfishNoteEncrypted,
  WasmNoteEncryptedHash,
  IronfishTransaction,
  SerializedWasmNoteEncrypted,
  SerializedWasmNoteEncryptedHash,
  SerializedTransaction
> {}

export type IronfishWitness = Witness<
  IronfishNoteEncrypted,
  WasmNoteEncryptedHash,
  SerializedWasmNoteEncrypted,
  SerializedWasmNoteEncryptedHash
>

export type IronfishMiningDirector = MiningDirector<
  IronfishNoteEncrypted,
  WasmNoteEncryptedHash,
  IronfishTransaction,
  SerializedWasmNoteEncrypted,
  SerializedWasmNoteEncryptedHash,
  SerializedTransaction
>

export type IronfishBlockSyncer = BlockSyncer<
  IronfishNoteEncrypted,
  WasmNoteEncryptedHash,
  IronfishTransaction,
  SerializedWasmNoteEncrypted,
  SerializedWasmNoteEncryptedHash,
  SerializedTransaction
>
