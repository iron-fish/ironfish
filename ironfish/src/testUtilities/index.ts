/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import './matchers'

export * from './helpers/serializable'
export * from './fixtures'
export * from './nodeTest'
export * from './utils'
export * from './witness'
