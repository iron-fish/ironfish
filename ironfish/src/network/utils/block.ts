/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
import bufio from 'bufio'
import { Assert } from '../../assert'
import { SerializedBlock } from '../../primitives/block'
import { GraffitiSerdeInstance } from '../../serde/serdeInstances'
import { BigIntUtils } from '../../utils/bigint'

export function writeBlock(
  bw: bufio.StaticWriter | bufio.BufferWriter,
  block: SerializedBlock,
): bufio.StaticWriter | bufio.BufferWriter {
  const { header, transactions } = block

  bw.writeU32(header.sequence)
  bw.writeHash(header.previousBlockHash)
  bw.writeHash(header.noteCommitment.commitment)
  bw.writeU32(header.noteCommitment.size)
  bw.writeHash(header.nullifierCommitment.commitment)
  bw.writeU32(header.nullifierCommitment.size)
  bw.writeBytes(BigIntUtils.toBytesLE(BigInt(header.target), 32))
  bw.writeBytes(BigIntUtils.toBytesLE(BigInt(header.randomness), 8))
  bw.writeU64(header.timestamp)

  Assert.isTrue(BigInt(header.minersFee) <= 0)
  bw.writeBytes(BigIntUtils.toBytesLE(-BigInt(header.minersFee), 8))

  bw.writeBytes(GraffitiSerdeInstance.deserialize(header.graffiti))

  bw.writeU16(transactions.length)
  for (const transaction of transactions) {
    bw.writeVarBytes(transaction)
  }

  return bw
}

export function readBlock(reader: bufio.BufferReader): SerializedBlock {
  const sequence = reader.readU32()
  const previousBlockHash = reader.readHash('hex')
  const noteCommitment = reader.readHash()
  const noteCommitmentSize = reader.readU32()
  const nullifierCommitment = reader.readHash('hex')
  const nullifierCommitmentSize = reader.readU32()
  const target = BigIntUtils.fromBytesLE(reader.readBytes(32)).toString()
  const randomness = BigIntUtils.fromBytesLE(reader.readBytes(8)).toString()
  const timestamp = reader.readU64()
  const minersFee = (-BigIntUtils.fromBytesLE(reader.readBytes(8))).toString()
  const graffiti = GraffitiSerdeInstance.serialize(reader.readBytes(32))

  const transactionsLength = reader.readU16()
  const transactions = []
  for (let j = 0; j < transactionsLength; j++) {
    transactions.push(reader.readVarBytes())
  }

  return {
    header: {
      sequence,
      previousBlockHash,
      noteCommitment: {
        commitment: noteCommitment,
        size: noteCommitmentSize,
      },
      nullifierCommitment: {
        commitment: nullifierCommitment,
        size: nullifierCommitmentSize,
      },
      target,
      randomness,
      timestamp,
      minersFee,
      graffiti,
    },
    transactions,
  }
}

export function getBlockSize(block: SerializedBlock): number {
  let size = 0
  size += 4 // sequence
  size += 32 // previousBlockHash
  size += 32 // noteCommitment.commitment
  size += 4 // noteCommitment.size
  size += 32 // nullifierCommitment.commitment
  size += 4 // nullifierCommitment.size
  size += 32 // target
  size += 8 // randomness
  size += 8 // timestamp
  size += 8 // minersFee
  size += 32 // graffiti

  size += 2 // transactions length
  for (const transaction of block.transactions) {
    size += bufio.sizeVarBytes(transaction)
  }
  return size
}
