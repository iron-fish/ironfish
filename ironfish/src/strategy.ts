/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import { Consensus } from './consensus'
import { Transaction } from './primitives/transaction'
import { MathUtils } from './utils'
import { WorkerPool } from './workerPool'

/**
 * Implementation of a Blockchain Strategy using zero-knowledge proofs.
 */
export class Strategy {
  readonly workerPool: WorkerPool
  readonly consensus: Consensus

  private miningRewardCachedByYear: Map<number, number>

  constructor(options: { workerPool: WorkerPool; consensus: Consensus }) {
    this.miningRewardCachedByYear = new Map<number, number>()
    this.workerPool = options.workerPool
    this.consensus = options.consensus
  }

  /**
   * Calculate the mining reward for a block based on its sequence
   *
   * See https://ironfish.network/docs/whitepaper/4_mining#include-the-miner-reward-based-on-coin-emission-schedule
   *
   * Annual coin issuance from mining goes down every year. Year is defined here by the
   * number of blocks
   *
   * Given the genesis block supply (genesisSupplyInIron) the formula to calculate
   * reward per block is:
   * (genesisSupply / 4) * e ^(-.05 * yearsAfterLaunch)
   * Where e is the natural number e (Euler's number), and -.05 is a decay function constant
   *
   * @param sequence Block sequence
   * @returns mining reward (in ORE) per block given the block sequence
   */
  miningReward(sequence: number): number {
    const ironFishYearInBlocks =
      (365 * 24 * 60 * 60) / this.consensus.parameters.targetBlockTimeInSeconds
    const yearsAfterLaunch = Math.floor(Number(sequence) / ironFishYearInBlocks)

    let reward = this.miningRewardCachedByYear.get(yearsAfterLaunch)
    if (reward) {
      return reward
    }

    const annualReward =
      (this.consensus.parameters.genesisSupplyInIron / 4) * Math.E ** (-0.05 * yearsAfterLaunch)

    reward = this.convertIronToOre(
      MathUtils.roundBy(annualReward / ironFishYearInBlocks, 0.125),
    )

    this.miningRewardCachedByYear.set(yearsAfterLaunch, reward)

    return reward
  }

  convertIronToOre(iron: number): number {
    return Math.round(iron * 10 ** 8)
  }

  /**
   * Create the miner's fee transaction for a given block.
   *
   * The miner's fee is a special transaction with one output and
   * zero spends. Its output value must be the total transaction fees
   * in the block plus the mining reward for the block.
   *
   * The mining reward may change over time, so we accept the block sequence
   * to calculate the mining reward from.
   *
   * @param totalTransactionFees is the sum of the transaction fees intended to go
   * in this block.
   * @param blockSequence the sequence of the block for which the miner's fee is being created
   * @param minerKey the spending key for the miner.
   */
  async createMinersFee(
    totalTransactionFees: bigint,
    blockSequence: number,
    minerSpendKey: string,
  ): Promise<Transaction> {
    // Create a new note with value equal to the inverse of the sum of the
    // transaction fees and the mining reward
    const amount = totalTransactionFees + BigInt(this.miningReward(blockSequence))

    const transactionVersion = this.consensus.getActiveTransactionVersion(blockSequence)

    return this.workerPool.createMinersFee(minerSpendKey, amount, '', transactionVersion)
  }
}
