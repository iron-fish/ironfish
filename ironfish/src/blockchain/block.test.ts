/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

import {
  TestStrategy,
  makeFakeBlock,
  blockHash,
  makeBlockWithTransaction,
  makeBlockAfter,
} from '../captain/testUtilities'
import { createNodeTest } from '../testUtilities/nodeTest'
import { BlockSerde } from './block'
import { useAccountFixture } from '../testUtilities/fixtures'
import { SerializedIronfishBlock } from '../strategy'

describe('Block', () => {
  const nodeTest = createNodeTest()

  it('correctly counts notes and nullifiers', async () => {
    await nodeTest.node.seed()

    const accountA = await useAccountFixture(nodeTest.node.accounts, 'accountA')
    const accountB = await useAccountFixture(nodeTest.node.accounts, 'accountB')
    const block = await makeBlockWithTransaction(nodeTest.node, accountA, accountB)

    expect(block.counts()).toMatchObject({
      nullifiers: 1,
      notes: 3,
    })

    const spends = Array.from(block.spends())
    expect(spends).toHaveLength(1)

    const notes = Array.from(block.allNotes())
    expect(notes).toHaveLength(3)
  }, 60000)

  it('serializes and deserializes a block', async () => {
    nodeTest.strategy.disableMiningReward()

    const genesis = await nodeTest.node.seed()
    const block = await makeBlockAfter(nodeTest.captain.chain, genesis)

    const serialized = nodeTest.captain.blockSerde.serialize(block)
    expect(serialized).toMatchObject({ header: { timestamp: expect.any(Number) } })

    const deserialized = nodeTest.captain.blockSerde.deserialize(serialized)
    expect(nodeTest.captain.blockSerde.equals(deserialized, block)).toBe(true)
  })

  it('throws when deserializing invalid block', () => {
    const serde = nodeTest.captain.blockSerde

    expect(() =>
      serde.deserialize(({ bad: 'data' } as unknown) as SerializedIronfishBlock),
    ).toThrowError('Unable to deserialize')
  })

  it('check block equality', () => {
    const strategy = new TestStrategy()
    const serde = new BlockSerde(strategy)

    const block1 = makeFakeBlock(strategy, blockHash(4), blockHash(5), 5, 5, 9)
    const block2 = makeFakeBlock(strategy, blockHash(4), blockHash(5), 5, 5, 9)

    block2.header.timestamp = block1.header.timestamp
    expect(serde.equals(block1, block2)).toBe(true)

    block2.header.randomness = 400
    expect(serde.equals(block1, block2)).toBe(false)

    const block3 = makeFakeBlock(strategy, blockHash(4), blockHash(5), 5, 5, 8)
    block3.header.timestamp = block1.header.timestamp
    expect(serde.equals(block1, block3)).toBe(false)

    const block4 = makeFakeBlock(strategy, blockHash(4), blockHash(5), 5, 5, 9)
    block4.header.timestamp = block1.header.timestamp
    block4.transactions[0].totalFees = BigInt(999)
    expect(serde.equals(block1, block4)).toBe(false)
  })
})
