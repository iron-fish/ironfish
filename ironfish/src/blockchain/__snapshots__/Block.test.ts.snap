// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block serializes and deserializes a block 1`] = `
Object {
  "header": Object {
    "count": 0,
    "graffiti": "66616B6520626C6F636B00000000000000000000000000000000000000000000",
    "graphId": 0,
    "hash": "0500000000000000000000000000000000000000000000000000000000000000",
    "isValid": false,
    "minersFee": "-14",
    "noteCommitment": Object {
      "commitment": "1-9",
      "size": 9,
    },
    "nullifierCommitment": Object {
      "commitment": "0000000000000000000000000000000000000000000000000000000000000000",
      "size": 1,
    },
    "previousBlockHash": "0400000000000000000000000000000000000000000000000000000000000000",
    "randomness": 0,
    "sequence": "5",
    "target": "115792089237316195423570985008687907853269984665640564039457584007913129639935",
    "timestamp": Any<Number>,
    "work": "0",
  },
  "transactions": Array [
    Object {
      "elements": Array [
        "5",
      ],
      "isValid": true,
      "spends": Array [],
      "totalFees": "1",
    },
    Object {
      "elements": Array [
        "6",
      ],
      "isValid": true,
      "spends": Array [],
      "totalFees": "1",
    },
    Object {
      "elements": Array [
        "7",
      ],
      "isValid": true,
      "spends": Array [],
      "totalFees": "1",
    },
    Object {
      "elements": Array [
        "8",
      ],
      "isValid": true,
      "spends": Array [],
      "totalFees": "1",
    },
    Object {
      "elements": Array [
        "9",
      ],
      "isValid": true,
      "spends": Array [],
      "totalFees": "-14",
    },
  ],
}
`;
